### PR TITLE
`sage.{interacts,misc,repl,typeset}`: Update `# needs`, modularization fixes

### PR DESCRIPTION
**Pull request by mkoeppe**
- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/misc/cython.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# sage.doctest: optional - sage.misc.cython
+# sage.doctest: needs sage.misc.cython
```
Comment (dcoudert) [inline src/sage/misc/cython.py:1]: It's surprising that we need this instruction to say that this file needs itself to be doctested

File: src/sage/misc/functional.py [hunk 2]
```diff
@@ -1836,43 +1878,43 @@ def squarefree_part(x):
...
+    Used internally to compute the square root of ``x``.
 
     INPUT:
 
-    -  ``x`` - a number
+    -  ``x`` -- a number
 
-    -  ``prec`` - None (default) or a positive integer
-       (bits of precision) If not None, then compute the square root
+    -  ``prec`` -- ``None`` (default) or a positive integer
```
Comment (dcoudert) [inline src/sage/misc/functional.py]: may be better: `a positive integer (default: ``None``); when specified, compute the square root with price bits of precision`

File: src/sage/misc/reset.pyx [hunk 3]
```diff
@@ -20,22 +20,22 @@ def reset(vars=None, attached=False):
...
 
-    Note that the variables in the set sage.misc.reset.EXCLUDE are
+    Note that the variables in the set :obj:`sage.misc.reset.EXCLUDE` are
     excluded from being reset.
 
     INPUT:
 
-    - ``vars`` - a list, or space or comma separated string (default:
+    - ``vars`` -- a list, or space or comma separated string (default:
       None), variables to restore
```
Comment (dcoudert) [inline src/sage/misc/reset.pyx]: ` ``None`` `

File: src/sage/misc/rest_index_of_methods.py [hunk 4]
```diff
@@ -255,12 +256,21 @@ def local_filter(f,name):
...
+
     functions = {getattr(root,name):name for name,f in root.__dict__.items() if
-                  (not name.startswith('_')          and # private functions
-                   not hasattr(f,'issue_number')      and # deprecated functions
-                   not inspect.isclass(f)            and # classes
-                   callable(getattr(f,'__func__',f)) and # e.g. GenericGraph.graphics_array_defaults
-                   local_filter(f,name))                 # possibly filter imported functions
+                  (not name.startswith('_')          and  # private functions
+                   can_import(f)                     and  # unresolved lazy imports
+                   not hasattr(f,'issue_number')     and  # deprecated functions
```
Comment (dcoudert) [inline src/sage/misc/rest_index_of_methods.py]: missing space `hasattr(f, 'issue_number')`

File: src/sage/misc/sageinspect.py [hunk 5]
```diff
@@ -105,10 +105,10 @@
 By :trac:`9976` and :trac:`14017`, introspection also works for interactively
 defined Cython code, and with rather tricky argument lines::
 
-    sage: cython('def foo(unsigned int x=1, a=\')"\', b={not (2+1==3):\'bar\'}, *args, **kwds): return')    # optional - sage.misc.cython
-    sage: print(sage_getsource(foo))                                                                        # optional - sage.misc.cython
+    sage: cython('def foo(unsigned int x=1, a=\')"\', b={not (2+1==3):\'bar\'}, *args, **kwds): return')                # needs sage.misc.cython
```
Comment (dcoudert) [inline src/sage/misc/sageinspect.py]: block scope to improve the alignment ?

File: src/sage/misc/sageinspect.py [hunk 6]
```diff
@@ -1504,36 +1507,43 @@ def sage_getargspec(obj):
 
     We now run sage_getargspec on some functions from the Sage library::
 
-        sage: sage_getargspec(identity_matrix)
-        FullArgSpec(args=['ring', 'n', 'sparse'], varargs=None, varkw=None, defaults=(0, False), kwonlyargs=[], kwonlydefaults=None, annotations={})
+        sage: sage_getargspec(identity_matrix)                                          # needs sage.modules
+        FullArgSpec(args=['ring', 'n', 'sparse'], varargs=None, varkw=None, defaults=(0, False),
+                    kwonlyargs=[], kwonlydefaults=None, annotations={})
         sage: sage_getargspec(factor)
         FullArgSpec(args=['n', 'proof', 'int_', 'algorithm', 'verbose'], varargs=None, varkw='kwds', defaults=(None, False, 'pari', 0), kwonlyargs=[], kwonlydefaults=None, annotations={})
```
Comment (dcoudert) [inline src/sage/misc/sageinspect.py]: multilines ?

File: src/sage/misc/table.py [hunk 7]
```diff
@@ -233,7 +240,7 @@ class table(SageObject):
 
     TESTS::
 
-        sage: TestSuite(table([["$x$", r"$\sin(x)$"]] +
+        sage: TestSuite(table([["$x$", r"$\sin(x)$"]] +                                             # needs sage.symbolic
```
Comment (dcoudert) [inline src/sage/misc/table.py]: alignement of `# needs` can be reduced

File: src/sage/misc/cython.py (hunk 1 above)
Comment (mkoeppe) [reply src/sage/misc/cython.py:1]: It is, but the feature "sage.misc.cython" also tests whether Cython is available and does a test compilation

**Comment by mkoeppe** [comment]
Thank you, all fixed.

**Review by dcoudert — APPROVED** [review]
LGTM.

**Comment by mkoeppe** [comment]
Thanks!

**Pull request merged**